File: shifai-ios/ShifAI/Presentation/Utilities/HapticFeedback.swift
import UIKit

/// Haptic Feedback Manager — provides tactile feedback for key interactions.
/// Centralizes all haptic patterns for consistency.
enum HapticFeedback {

    // MARK: - Standard Patterns

    /// Light tap — tab switch, toggle
    static func light() {
        UIImpactFeedbackGenerator(style: .light).impactOccurred()
    }

    /// Medium tap — save, confirm
    static func medium() {
        UIImpactFeedbackGenerator(style: .medium).impactOccurred()
    }

    /// Heavy tap — delete, important action
    static func heavy() {
        UIImpactFeedbackGenerator(style: .heavy).impactOccurred()
    }

    /// Success — save completed, sync done
    static func success() {
        UINotificationFeedbackGenerator().notificationOccurred(.success)
    }

    /// Warning — approaching limit, quiet hours
    static func warning() {
        UINotificationFeedbackGenerator().notificationOccurred(.warning)
    }

    /// Error — validation failed, sync error
    static func error() {
        UINotificationFeedbackGenerator().notificationOccurred(.error)
    }

    /// Selection changed — slider, picker
    static func selection() {
        UISelectionFeedbackGenerator().selectionChanged()
    }

    // MARK: - App-Specific Patterns

    /// Daily log saved successfully
    static func dailyLogSaved() {
        success()
    }

    /// Symptom added to body map
    static func symptomAdded() {
        medium()
    }

    /// Flow slider changed
    static func sliderChanged() {
        selection()
    }

    /// Account deletion confirmed
    static func destructiveAction() {
        heavy()
    }

    /// Biometric unlock success
    static func biometricSuccess() {
        success()
    }
}

File: shifai-android/app/src/main/java/com/shifai/presentation/util/HapticFeedback.kt
package com.shifai.presentation.util

import android.content.Context
import android.os.Build
import android.os.VibrationEffect
import android.os.Vibrator
import android.os.VibratorManager

/**
 * Haptic Feedback — tactile feedback for key interactions.
 * Mirrors iOS HapticFeedback.swift.
 */
object HapticFeedback {

    // ─── Standard Patterns ───

    /** Light tap — tab switch, toggle */
    fun light(context: Context) = vibrate(context, 20, 80)

    /** Medium tap — save, confirm */
    fun medium(context: Context) = vibrate(context, 40, 128)

    /** Heavy tap — delete, important action */
    fun heavy(context: Context) = vibrate(context, 60, 200)

    /** Success — save completed, sync done */
    fun success(context: Context) = vibrate(context, 30, 100)

    /** Warning — approaching limit */
    fun warning(context: Context) = vibrate(context, 50, 150)

    /** Error — validation failed */
    fun error(context: Context) = vibrate(context, 70, 200)

    /** Selection changed — slider, picker */
    fun selection(context: Context) = vibrate(context, 10, 60)

    // ─── App-Specific Patterns ───

    fun dailyLogSaved(context: Context) = success(context)
    fun symptomAdded(context: Context) = medium(context)
    fun sliderChanged(context: Context) = selection(context)
    fun destructiveAction(context: Context) = heavy(context)
    fun biometricSuccess(context: Context) = success(context)

    // ─── Implementation ───

    private fun vibrate(context: Context, durationMs: Long, amplitude: Int) {
        val vibrator = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            val vm = context.getSystemService(Context.VIBRATOR_MANAGER_SERVICE) as VibratorManager
            vm.defaultVibrator
        } else {
            @Suppress("DEPRECATION")
            context.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            vibrator.vibrate(VibrationEffect.createOneShot(durationMs, amplitude))
        } else {
            @Suppress("DEPRECATION")
            vibrator.vibrate(durationMs)
        }
    }
}
